src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::fetchMetadataAction()
+{
+    requireParams({u"source"_s});
+
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    // must provide some value to parse
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QUrl sourceUrl = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+    const QString source = sourceUrl.toString();
+
+    // metadata has already been requested
+    if (const auto iter = m_torrentSource.constFind(source); iter != m_torrentSource.constEnd())
+    {
+        const BitTorrent::InfoHash infoHash = iter.value();
+        // metadata has already been downloaded
+        if (isMetadataDownloaded(infoHash))
+        {
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+        }
+        else
+        {
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
+        }
+
+        return;
+    }
+
+    // http(s) url
+    if (Net::DownloadManager::hasSupportedScheme(source))
+    {
+        qDebug("Fetching torrent %s", qUtf8Printable(source));
+        const auto *pref = Preferences::instance();
+
+        Net::DownloadManager::instance()->download(Net::DownloadRequest(source).limit(pref->getTorrentFileSizeLimit())
+                , pref->useProxyForGeneralPurposes(), this, &TorrentsController::onDownloadFinished);
+
+        m_torrentSource.insert(source, {});
+
+        setStatus(APIStatus::Async);
+        setResult(QJsonObject {});
+        return;
+    }
+
+    // magnet link or info hash
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+        auto *session = BitTorrent::Session::instance();
+        // torrent already exists in transfer list
+        if (const BitTorrent::Torrent *const torrent = session->findTorrent(infoHash); torrent)
+        {
+            // torrent's metadata may not have downloaded yet
+            if (torrent->info().isValid())
+            {
+                setResult(serializeTorrentInfo(*torrent));
```

Comment:
What actions do you expect the client to take in this case? Won't it just try to add the same torrent again?

Reply:
The code comment is a bit confusing in its current location. In the highlighted case, the full torrent info is returned. In the `else` case, we return partial information (torrent hash + id). This is purely informational for the caller and lets them know via the 202 that that full metadata isn't available yet.

In my usage of this API in #21645, I display the partial returned information but continue to show the "Retrieving metadata" spinner. This matches the behavior of the GUI - it immediately shows the info hash while loading other metadata in the background.

Reply:
I mean the case when there is existing torrent in qBittorrent. You just send parsed metadata in regular way. So my questions above.

Reply:
It's up to the client to determine how they want to handle that. This endpoint just returns metadata about a torrent, so I don't think its response should be any different than it is now. The most likely case is that the client will try to add the torrent and then get an appropriate error.